src/SessionData.php

```diff
@@ -39,23 +40,31 @@ class SessionData implements Session
     private $is_new = false;
 
     /**
-     * @param string $session_id
+     * @param string $client_session_id
      * @param array  $data
      * @param bool   $is_new
      */
-    public function __construct(string $session_id, array $data, bool $is_new)
+    public function __construct(string $client_session_id, array $data, bool $is_new)
     {
-        $this->session_id = $session_id;
+        $this->client_session_id = $client_session_id;
```

Comment:
I would much prefer if the client_session_id was not a part of the SessionData object at all. It is only needed for generating the stored session_id (at session resumption time) and generating the session cookie (at session creation time), after all.

With the proposed code it is very easy for a module to grab the SessionData object and pull out the client_session_id (out of ignorance or in an attempt to be "clever").

The only part of the code that relies on having the client_session_id available (as far as I can tell) is commitSession(), but since we've decided to treat this release as a breaking change anyway, we have the freedom to tweak the API to get rid of this need.

Reply:
@rhl-jfm I tried, and this seems substantially too complex for the limited value it adds.

I attempted a simple solution where the Session Service would maintain a map (where Session ID => Client Session ID) but this doesn't address the case for `Session::renew()` which needs to be able to create a new, valid Session ID.

So I could make that method public on the Session Service, but then we'd just have a new version of the same problem.

So I attempted a more complex solution, introducing a a `SessionRegistry` that creates/resolves Client Session IDs - the Session Service creates and keeps and instance of this privately, and issues it to `SessionData` instances via the constructor, allowing them to access the registry for Session ID renewal.

This works, but creates new problems, since now we're collecting Session IDs in this registry, which means memory-leaks under long-running apps - and I can add more work-arounds to address that: clean them up in the `SessionData` destructor, be extremely strict about registering/creating the IDs, since their lifecycle needs to exactly match that of `SessionData` and so on.

Even with all that in place (a new dependecy with 100 lines of code, and I haven't started refactoring the tests yet) and relative certainty of no memory-leaks, it occurs to me, this will actually break nested/recursive requests (e.g. soft redirects) through the middleware stack, since `SessionData::__destruct` will remove the Session ID from the registry, so a nested request would trigger an exception.

There's no work-around for that. (and no, we don't perform nested requests at the moment, but I don't want a nasty surprise the day somebody decides to try that.)

I can push the work I've done so far, if you'd like to review it. Maybe you can see a different way - and I'd invite you to submit a PR, but I'm betting you don't have time for that, so.

I'd like to propose we keep it as is for this release, and, in addition, we can mark the `getClientSessionID()` method as `@internal`.

Note that `SessionData::getClientSessionID()` isn't part of the public `Session` interface, so this method is only visible to you in the first place if you're using the wrong type-hints.

Reply:
I accept your explanation :smiley:

Given that the method is not part of the public `Session` interface, you can't really use it by accident. And if you really want to get at the raw cookie value you can always look at $_COOKIE anyway...

Reply:
I am just wondering, could we benefit from this?

`variables_order string`
> Sets the order of the EGPCS (Environment, Get, Post, Cookie, and Server) variable parsing. For example, if variables_order is set to "SP" then PHP will create the superglobals $_SERVER and $_POST, but not create $_ENV, $_GET, and $_COOKIE. __Setting to "" means no superglobals will be set.__

We may use zends register from globals method right now, and I don't know if we could map it to ServerRequest without the superglobals, we need to pass the headers and payload on our own, `php://input`?

Reply:
@boan2010 no, `php://input` only gives you the body of the request - the headers are only available to your in a semi-mangled form via `$_SERVER`. PHP runs under CGI, not under HTTP, unfortunately - packages like zend-diactoros have to work backwards from CGI values to something resembling the original HTTP request.

Reply:
Yep, but maybe we could get the headers from something like this: http://php.net/manual/en/function.getallheaders.php without using the superglobals... But this is with apache, don't know if you could do it otherwise... 

> This function became available under FastCG (?)

Reply:
@boan2010 `getallheaders()` is just another way to depend on global state - and, as you pointed out, only works with Apache. A lot of the "win" with PSR-7/15 comes from never relying on global state.

(is this discussion in any way related to Rene's comment, or is this a complete side-track? I'm happy to discuss in-person 😄)

Reply:
Just a note from our talk :-)

https://github.com/Nyholm/psr7-server/blob/master/src/ServerRequestCreator.php#L52